pkg/trimaran/lowriskovercommitment/lowriskovercommitment.go

```diff
@@ -0,0 +1,308 @@
+	return nil
+}
+
+// computeRank : rank function for the LowRiskOverCommitment
+func (pl *LowRiskOverCommitment) computeRank(metrics []watcher.Metric, nodeInfo *framework.NodeInfo, pod *v1.Pod,
+	podRequests *framework.Resource, podLimits *framework.Resource) float64 {
+	// get all pods on node
+	node := nodeInfo.Node()
+	nodeName := node.GetName()
+	var podsOnNode []*v1.Pod
+	for _, podInfo := range nodeInfo.Pods {
+		if podInfo != nil && podInfo.Pod != nil {
+			podsOnNode = append(podsOnNode, podInfo.Pod)
+		}
+	}
+	// calculate risk based on requests and limits
+	nodeRequestsAndLimits := trimaran.GetNodeRequestsAndLimits(podsOnNode, node, pod, podRequests, podLimits)
+	riskCPU := pl.computeRisk(metrics, v1.ResourceCPU, watcher.CPU, node, nodeRequestsAndLimits)
+	riskMemory := pl.computeRisk(metrics, v1.ResourceMemory, watcher.Memory, node, nodeRequestsAndLimits)
+	rank := 1 - math.Max(riskCPU, riskMemory)
+
+	klog.V(6).InfoS("Node rank", "nodeName", nodeName, "riskCPU", riskCPU, "riskMemory", riskMemory, "rank", rank)
+
+	return rank
+}
+
+// computeRisk : calculate the risk of scheduling on node for a given resource
+func (pl *LowRiskOverCommitment) computeRisk(metrics []watcher.Metric, resourceName v1.ResourceName,
+	resourceType string, node *v1.Node, nodeRequestsAndLimits *trimaran.NodeRequestsAndLimits) float64 {
+	var riskLimit, riskLoad, totalRisk float64
+
+	defer func() {
+		klog.V(6).InfoS("Calculated risk", "node", klog.KObj(node), "resource", resourceName,
+			"riskLimit", riskLimit, "riskLoad", riskLoad, "totalRisk", totalRisk)
+	}()
+
+	nodeRequest := nodeRequestsAndLimits.NodeRequest
+	nodeLimit := nodeRequestsAndLimits.NodeLimit
+	nodeRequestMinusPod := nodeRequestsAndLimits.NodeRequestMinusPod
+	nodeLimitMinusPod := nodeRequestsAndLimits.NodeLimitMinusPod
+	nodeCapacity := nodeRequestsAndLimits.Nodecapacity
+
+	var request, limit, capacity, requestMinusPod, limitMinusPod int64
+	if resourceName == v1.ResourceCPU {
+		request = nodeRequest.MilliCPU
+		limit = nodeLimit.MilliCPU
+		requestMinusPod = nodeRequestMinusPod.MilliCPU
+		limitMinusPod = nodeLimitMinusPod.MilliCPU
+		capacity = nodeCapacity.MilliCPU
+	} else if resourceName == v1.ResourceMemory {
+		request = nodeRequest.Memory
+		limit = nodeLimit.Memory
+		requestMinusPod = nodeRequestMinusPod.Memory
+		limitMinusPod = nodeLimitMinusPod.Memory
+		capacity = nodeCapacity.Memory
+	} else {
+		// invalid resource
+		klog.V(6).InfoS("Unexpected resource", "resourceName", resourceName)
+		return 0
+	}
+
+	// (1) riskLimit : calculate overcommit potential load
+	excess := limit - request
+	// This should not happen
+	if excess < 0 {
+		klog.V(6).InfoS("Negative excess", "node", klog.KObj(node), "resource", resourceName,
+			"request", request, "limit", limit)
+		excess = 0
+	}
+	maxLoad := capacity
+	if limit < capacity {
+		maxLoad = limit
+	}
+	allowed := maxLoad - request
+	if allowed < 0 {
+		allowed = 0
+	}
```

Comment:
suppose request/limit/capacity as 1/1/10, then allowed=0, it doesn't look right to me.

Reply:
I still feel this section is hard to read and can be simplified:

- if limits <= capacity, it seems we can just claim riskLimit = 0, correct? because in this case, allowed === excess
- if limits > capacity, some logic is valid, but what if limits equals requests, like request/limits/capacity as 15/15/10, it seems in this case, riskLimit is also 0?

Reply:
@Huang-Wei Agreed. My proposal for simplification is to replace
```
// (1) riskLimit : calculate overcommit potential load
	excess := limit - request
	// This should not happen
	if excess < 0 {
		klog.V(6).InfoS("Negative excess", "node", klog.KObj(node), "resource", resourceName,
			"request", request, "limit", limit)
		excess = 0
	}
	maxLoad := capacity
	if limit < capacity {
		maxLoad = limit
	}
	allowed := maxLoad - request
	if allowed < 0 {
		allowed = 0
	}
	if excess > 0 && allowed < excess {
		riskLimit = 1 - float64(allowed)/float64(excess)
	}
```
with
```
// (1) riskLimit : calculate overcommit potential load
	if limit > capacity {
		riskLimit = float64(limit - capacity)/float64(limit - request)
	}
```

The reasons I believe it should work are:

- `riskLimit` is initialized to zero, so the case of `limit <= capacity` is handled automatically
- conditions `request <= limit` and `request <= capacity` are already ensured by `GetNodeRequestsAndLimits()`
- denominator cannot be zero since `limit > capacity` and `request <= capacity`
- if `limit == request` then it must be that `limit <= capacity`, hence `riskLimit` will remain zero

One interpretation of the simplified equation is: the numerator captures the potential of bursting beyond capacity (overload), the denominator captures the potential of bursting beyond allocation, and the ratio is the chance of overload when bursting (between 0 and 1).

Now, we come to the case "requests/limits/capacity as 15/15/10". This case will not arise since requests (15) cannot be larger than capacity (10). (The pods will not fit on the node.) If that was 8/8/10, then the risk is 0.

I will make the change, unless you have other suggestions. Thanks.